contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,439 @@
+selectable with a low probability of overlap.
+    - ControllerRefs will be used as the authoritative test for ownership.
+    - The specification type Object's `.Selector` should be used where 
+    applicable. 
+    - Alternatively, a Kind unique label may be set to the `.Name` of the 
+    specification type Object.
+1. The controller will add a ControllerRef indicating the specification type 
+Object as the owner of the ControllerRevision in the ControllerRevision's 
+`.OwnerReferences`.
+1. The controller will use the hash from above, along with a user identifiable 
+prefix, to [generate a unique `.Name`](#unique-name-generation) for the 
+ControllerRevision.
+    - The controller should, where possible, use the `.Name` of the 
+    specification type Object.
+1. The controller will persist the ControllerRevision via the API Server.
+    - Note that, in practice, creation occurs concurrently with 
+    [collision resolution](#collision-resolution).
+  
+### Revision Number Selection
+We propose two methods for selecting the `.Revision` used to order a 
+specification type Object's revision history. 
+
+1. Set the `.Revision` field to the `.Generation` field. 
+    - This approach has the benefit of leveraging the existing monotonically 
+    increasing sequence generated by `.Generation` field. 
+    - The downside of this approach is that history will not survive the 
+    destruction of an Object.
+1. Use an approach analogous to Deployment. 
+    1. Reconstruct the Object's revision history.
+    1. If the history is empty, use a `.Revision` of `0`.
+    1. If the history is not empty, set history to a value greater than the 
+    maximum value of all previous `.Revisions`.
+
+### History Reconstruction
+To reconstruct the history of a specification type Object, a controller will do 
+the following. 
+
+1. Select all ControllerRevision Objects labeled as described 
+[above](#version-snapshot-creation).
+1. Filter any ControllerRevisions that do not have a ControllerRef in their 
+`.OwnerReferences` indicating ownership by the Object.
+1. Sort the ControllerRevisions by the `.Revision` field.
+1. This produces a strictly ordered set of ControllerRevisions that comprises 
+the ordered revision history of the specification type Object.
+
+### History Maintenance
+Controllers should be configured, either globally or on a per specification type 
+Object basis, to have a `RevisionHistoryLimit`. This field will indicate the 
+number of non-live revisions the controller should maintain in its history
+for each specification type Object. Every time a controller observes a 
+specification type Object it will do the following.
+
+1. The controller will 
+[reconstruct the Object's revision history](#history-reconstruction).
+    - Note that the process of reconstructing the Object's history filters any
+    ControllerRevisions not owned by the Object.
+1. The controller will filter any ControllerRevisions that represent a live 
```

Comment:
No. When we do equality detection we will relabel products as necessary and not create a new live version where this is not necessary.